src/staging-heap.cpp

```diff
@@ -0,0 +1,253 @@
+    SLANG_RHI_ASSERT(m_pages.size() == 0);
+    m_pages.clear();
+}
+
+RefPtr<StagingHeap::Handle> StagingHeap::allocHandle(size_t size, MetaData metadata)
+{
+    Allocation allocation = alloc(size, metadata);
+    return new Handle(this, allocation);
+}
+
+StagingHeap::Allocation StagingHeap::alloc(size_t size, MetaData metadata)
+{
+    // Get aligned size.
+    size_t aligned_size = alignUp(size);
+
+    // Attempt to allocate from page if size is less than page size.
+    if (aligned_size < m_page_size)
+    {
+        for (auto& page : m_pages)
+        {
+            std::list<Node>::iterator node;
+            if (page.second->allocNode(aligned_size, metadata, node))
+            {
+                Allocation res;
+                res.buffer = page.second->getBuffer();
+                res.node = node;
+                m_total_used += aligned_size;
+                return res;
+            }
+        }
+    }
+
+    // Can't fit in existing page, so allocate from new one
+    size_t page_size = aligned_size < m_page_size ? m_page_size : aligned_size;
+    RefPtr<Page> page = allocPage(page_size);
+    std::list<Node>::iterator node;
+    page->allocNode(aligned_size, metadata, node);
+
+    Allocation res;
+    res.buffer = page->getBuffer();
+    res.node = node;
+    m_total_used += aligned_size;
+
+    return res;
+}
+
+void StagingHeap::free(Allocation allocation)
+{
+    // Decrement overall total used (before freeing node).
+    m_total_used -= allocation.node->size;
+
+    // Free the node from the page.
+    RefPtr<Page> page = m_pages[allocation.getPageId()];
+    page->freeNode(allocation.node);
+
+    // Free page if now have more than 1 empty page or this is none-standard page size.
+    if (page->getUsed() == 0)
+    {
+        if (page->getCapacity() == m_page_size)
+        {
+            int empty_pages = 0;
+            for (auto& p : m_pages)
+                if (p.second->getUsed() == 0)
+                    empty_pages++;
+            if (empty_pages > 1)
+                m_pages.erase(page->getId());
+        }
+        else
+        {
+            m_pages.erase(page->getId());
+        }
+    }
+}
+
+RefPtr<StagingHeap::Page> StagingHeap::allocPage(size_t size)
+{
+    ComPtr<IBuffer> bufferPtr;
+    BufferDesc bufferDesc;
+    bufferDesc.usage = BufferUsage::CopyDestination | BufferUsage::CopySource | BufferUsage::ConstantBuffer;
+    bufferDesc.defaultState = ResourceState::General;
+    bufferDesc.memoryType = MemoryType::Upload;
+    bufferDesc.size = size;
+
+    // TODO(staging-heap): Check for failure
+    m_device->createBuffer(bufferDesc, nullptr, bufferPtr.writeRef());
+
+    // Create page and store buffer pointer
+    RefPtr<Page> page = new Page(m_next_page_id++, checked_cast<Buffer*>(bufferPtr.get()));
+    m_pages.insert({page->getId(), page});
+    m_total_capacity += size;
+
+    // TODO(staging-heap): Find out what this is
+    //  The buffer is owned by the page.
+    page->getBuffer()->comFree();
```

Comment:
this is ugly, but basically we tell the buffer to break the strong reference to `Device` and avoid a reference cycle